### ISSUE TITLE
Is -charge1*charge2*erf(alphaR) sufficient to cancel reciprocal contribution?

### ISSUE THREAD

**Issue by udlich** [openmm/openmm#3526]
Hello everyone, I want to do some alchemical issues, and I need to make sure that atom1 and atom2 never see each other. I use NonbondedForce with Ewald method, and set exceptions between atom1 and atom2.

However, I feel confused about how to exactly cancel the electrostatic interaction between atom1 and atom2. I know NonbondedForce excludes atom1-atom2 pair in the direct space calculation step, and minus charge1*charge2*erf(alphaR) to cancel the reciprocal contribution of thie pair.

I think charge1*charge2*erf(alphaR) only cancels the pair atom1(lattice 0)-atom2(lattice 0), and won't cancel the many pairs atom1(lattice 0) - atom2(lattice 1, 2...). 

Would you please tell what mistake I made? Thank you!

**Comment by peastman** [comment]
That's correct.  Nonbonded exceptions are typically used for particles that are bonded to each other (1-2, 1-3, and 1-4 interactions).  When using periodic boundary conditions, each "particle" represents an infinite grid of particles repeating through space.  The exception alters the interaction between two particles that are bonded to each other, but not the interaction of each one with all the other periodic copies of the other.

I don't believe there's any way to remove the interactions between all possible copies of two particles.  This is inherent in the way PME works.  It first spreads all the charges onto a grid, then computes the interaction of each grid point with every other.  It doesn't keep track of which particles the charge came from.  It just computes how the charge distribution interacts with itself.

**Comment by udlich** [comment]
Thanks for your reply! I concern on this because in alchemistry there are many cases to cancel the interaction between two atoms. For example, unique atoms of the initial state should not see any unique atoms of the final state.Therefore, my question is not subjected to bonded 1-2, 1-3 and 1-4 pairs.

Is there any chance to insert some codes in the inner k iteration part of the reciprocal section, after line 357. I give a python-like pseudo code below, and could you please assess such a modification?

349                    for (int n = 0; n < numberOfAtoms; n++) {
350                        double force = ak * (cs * tab_qxyz[n].imag() - ss * tab_qxyz[n].real());
351                        forces[n][0] += 2 * recipCoeff * force * kx ;
352                        forces[n][1] += 2 * recipCoeff * force * ky ;
353                        forces[n][2] += 2 * recipCoeff * force * kz ;
354                    }
355
356                    recipEnergy       = recipCoeff * ak * (cs * cs + ss * ss);
357                    totalRecipEnergy += recipEnergy;

newline1            for pair in exclusionlist:
newline2                 atom1, atom2 = pair            # then get the atom charges
newline3                 r = sqrt((x12+kx)**2+(y12+ky)**2+(z12+kz)**2)      # we are in the inner k-iteration
newline4                 recipEnergy -= charge1 * charge2 * erf(alpha * r) / r    # each k needs to minus such an energy term

If I miss or misunderstand anything, can you point it out? Thank you!

**Comment by peastman** [comment]
The code you're looking at is for conventional Ewald summation, not PME.  And it doesn't do what you think it does.  In that case the information is encoded in coefficients for basis functions rather than a grid.  But the principle is the same: build up a representation of the charge distribution, then compute how the particles interact with that distribution.

**Comment by udlich** [comment]
Many thanks for your explanations!
